components/in_memory_engine/src/region_stats.rs

```diff
@@ -269,68 +215,96 @@ impl RegionStatsManager {
-                        // does not evict it this time.
-                        let time_loaded = regions_loaded.entry(r.id).or_insert(Instant::now());
-                        if Instant::now() - *time_loaded < self.evict_min_duration {
-                            None
-                        } else {
-                            Some((r, s))
-                        }
-                    })
-                    .rev() // evict the regions with least next + prev after the filter
-                    .take(max_count_to_evict)
-                    .collect();
+        if cached_region_stats.len() <= MIN_REGION_COUNT_TO_EVICT {
+            return (regions_to_load, vec![]);
+        }
 
-            // TODO(SpadeA): remove it after it's stable
-            let debug: Vec<_> = region_to_evict
-                .iter()
-                .map(|(r, s)| {
-                    format!(
-                        "region_id={}, cop={}, cop_detail={:?}, mvcc_amplification={}",
-                        r.id,
-                        s.query_stats.coprocessor,
-                        s.cop_detail,
-                        s.cop_detail.mvcc_amplification(),
-                    )
-                })
-                .collect();
-            info!(
-                "ime collect regions to evict";
-                "reached_stop_limit" => reach_stop_load,
-                "regions" => ?debug,
-            );
-            region_to_evict.into_iter().map(|(r, _)| r).collect()
-        } else {
-            vec![]
-        };
+        let mut total_next_prev = 0;
+        let mut total_cop_requests = 0;
+        for crs in &cached_region_stats {
+            total_next_prev += crs.stat.cop_detail.iterated_count();
+            total_cop_requests += crs.sma_cop_requests_avg;
+        }
+        let avg_next_prev = total_next_prev / cached_region_stats.len();
+        let avg_cop_requests = total_cop_requests / cached_region_stats.len();
+        let is_memory_low = memory_controller.reached_stop_load_threshold();
+
+        let now = Instant::now();
+        let mut region_to_evict = Vec::with_capacity(max_count_to_evict);
+        for crs in cached_region_stats.into_iter().take(max_count_to_evict) {
+            let need_evict = {
+                // In case, memory usage is relatively low, we evict those that
+                // should not be cached.
+                //
+                // NB: When a region is cached, its RegionWriteCfCopDetail only
+                // reflects the MVCC amplification in IME not in RocksDB. So
+                // low MVCC amplification is not a good indicator for eviction.
+                //
+                // Instead, we assume workload patterns remain consistent after
+                // caching and evict based on request frequency.
+                let is_cop_requests_low = crs.sma_cop_requests_avg
+                    <= avg_cop_requests / SMA_COP_REQUEST_AVG_FILTER_FACTOR;
+
+                let is_cop_requests_reliable =
```

Comment:
Does this mean if a region has no reuqest, it can be evicted at all?

Reply:
`is_cop_requests_low` means if a region has a few requests, then it will be an eviction candidate.
`is_cop_requests_reliable` means an eviction candidate can only be evicted if IME has sampled it three times (by default 30 minutes)

Reply:
Could you update the annotations with the above comments or incorporate them into the annotations? This would make the strategy for choosing eviction candidates clearer and easier to understand.

Reply:
Comments are added in f263f5e07f98da24d1fc65a9f694693ac5835f30